crates/subspace-networking/src/node_runner.rs

```diff
@@ -1430,10 +1430,26 @@ where
             }
             Command::GenericRequest {
                 peer_id,
+                addresses,
                 protocol_name,
                 request,
                 result_sender,
             } => {
+                if !addresses.is_empty()
+                    && !self
+                        .swarm
+                        .connected_peers()
+                        .any(|candidate| candidate == &peer_id)
+                {
+                    if let Err(error) = self.swarm.dial(
```

Comment:
Since we get addresses through out request/response protocol, libp2p will not know where to dial if we just use peer ID, so we need to do a bit of work ourselves

Reply:
The libp2p-friendly approach would be to use `extend_addresses_through_behaviour` feature. We never used it before though. How does a separate async `dial` call work with the following request? Won't we have the undefined behavior case because of the unsynchronized connection requests and an actual request?

Reply:
I'm not sure how `extend_addresses_through_behaviour` will help us if none of the behaviors know about the address anyway. Are you suggesting to create another behavior just to store addresses (for how long?) so we can use `extend_addresses_through_behaviour` instead of providing addresses explicitly? That doesn't seem better to me.

As to the race, I assume it'll wait for the dial to finish, but without integration into other things I was not able to test it. I will later though and correct this if it doesn't work as expected.

Reply:
I did not. I find dealing with libp2p (at least the way it is implemented right now) quite unpleasant, so if I can achieve something without writing behaviors I prefer doing that.

I agree it is somewhat complex, but it IMO is still easier to understand than libp2p-kad's query implementation, despite what I wrote is probably doing significantly more work in practice.

In ideal world dedicated behavior would probably be preferred, but as it stands today, I strongly prefer our request/response infrastructure.